### PR TITLE
Feature/attach to vpc sc perimeter

### PR DESCRIPTION
**Pull request by kwraith05**
Adding ability to add project to an existing VPC service control perimeter (service perimeter attachment)

@bharathkkb  Please review and let me know of any changes I should make

### CODE REVIEW

**Comment by bharathkkb** [comment]
@kwraith05 seems like there is some API enablement missing from the setup?

**Comment by kwraith05** [comment]
@morgante  the logs for the build say the creation of the accesscontextmanager api project service was completed successfully, meaning the api should be enabled. However, the error says that if the API was enabled, it could take a few minutes for the action to propagate in the system and to wait a few minutes before trying again. 
@bharathkkb and I discussed and we're thinking we're getting this error because the project the master SA resides in doesn't have the API enabled. Do you know of any workaround?

**Comment by bharathkkb** [comment]
Also @morgante for some additional context, the reason for moving `access_context_manager_policy` into the `setup` is because @kwraith05 ran into an issue with the `count value depends on resource attributes` error.

**Comment by morgante** [comment]
> @bharathkkb and I discussed and we're thinking we're getting this error because the project the master SA resides in doesn't have the API enabled. Do you know of any workaround?

That's probably be the issue.

> Also @morgante for some additional context, the reason for moving access_context_manager_policy into the setup is because @kwraith05 ran into an issue with the count value depends on resource attributes error.

I don't quite follow. Why would we need to create it in `setup`? We should track down the actual reason for that error instead of moving it to the setup unnecessarily.

**Comment by bharathkkb** [comment]
@morgante I believe the error originated from 
https://github.com/terraform-google-modules/terraform-google-project-factory/blob/5a2be35a67014e8ff50682dcecf017e014301067/modules/core_project_factory/main.tf#L68

because value `var. vpc_service_control_perimeter_name` is not known until we actually create it. @kwraith05 could you confirm the issue and post logs if any?

**Comment by morgante** [comment]
@bharathkkb If we're encountering that issue, I suggest exposing the parameter separately (`vpc_service_control_attach_enabled`) instead of trying to compute it. If we're seeing it, users will too.

**Comment by bharathkkb** [comment]
Makes sense. @kwraith05 let me know if you need any help with the change recommended above.

**Comment by bharathkkb** [comment]
@kwraith05 I think that error is stemming from https://github.com/terraform-providers/terraform-provider-google/issues/3582. As a fix can you bump it to latest wherever possible?

File: variables.tf [hunk 1]
```diff
@@ -208,3 +208,9 @@ variable "skip_gcloud_download" {
   type        = bool
   default     = false
 }
+
```
Comment (bharathkkb) [inline variables.tf:211]: Wouldnt we need to expose `vpc_service_control_attach_enabled` as well and pass it in to core?

**Comment by bharathkkb** [comment]
Also @kwraith05 I think the tests did not actually invoke the right code path as `vpc_service_control_attach_enabled` is set to `false` by default. In the fixture we need to set this to `true` after you do https://github.com/terraform-google-modules/terraform-google-project-factory/pull/428#discussion_r455983383

File: test/fixtures/vpc_sc_project/variables.tf [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+variable "parent_id" {
+  type        = string
+  description = "The parent of this AccessPolicy in the Cloud Resource Hierarchy. As of now, only organizations are accepted as parent."
+  default     = "11111111"
```
Comment (morgante) [inline test/fixtures/vpc_sc_project/variables.tf:20]: This is likely redundant with org_id? Let's remove if we don't need it.

File: test/fixtures/vpc_sc_project/variables.tf [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+variable "policy_name" {
+  type        = string
+  description = "The policy's name"
+  default     = "policy1"
```
Comment (morgante) [inline test/fixtures/vpc_sc_project/variables.tf:26]: Remove this as it's not a sensible default.

File: test/fixtures/vpc_sc_project/variables.tf [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  type        = string
+  description = "Billing account ID"
+}
+
+variable "random_string_for_testing" {
+  type        = string
+  description = "A random string of characters to be appended to resource names to ensure uniqueness"
+}
+
+variable "vpc_service_control_attach_enabled" {
```
Comment (morgante) [inline test/fixtures/vpc_sc_project/variables.tf:49]: This doesn't need to be a variable. The whole point of this fixture is to test VPCSC so we should hardcode to true.

File: test/fixtures/vpc_sc_project/variables.tf [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  type        = bool
+  description = "Whether the project will be attached to a VPC Service Control Perimeter"
+  default     = false
+}
+
+variable "vpc_service_control_perimeter_name" {
```
Comment (morgante) [inline test/fixtures/vpc_sc_project/variables.tf:55]: Is this needed? We don't use it in the test.

File: test/fixtures/vpc_sc_project/variables.tf [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  description = "The name of an existing VPC Service Control Perimeter to add the created project to"
+  default     = null
+}
+
+variable "policy_id" {
+  type        = string
+  description = "The ID of the access context manager policy the perimeter lies in"
+}
+
+variable "service_account_email" {
```
Comment (morgante) [inline test/fixtures/vpc_sc_project/variables.tf:66]: Don't make this configurable, just leave sa_email off entirely.

File: test/integration/vpc_sc_project/controls/vpc_sc_project.rb [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  describe command("gcloud services list --project #{project_id}") do
+    its('exit_status') { should be 0 }
+    its('stderr') { should eq '' }
+
+    its('stdout') { should match(/compute\.googleapis\.com/) }
+    its('stdout') { should match(/accesscontextmanager\.googleapis\.com/) }
+    its('stdout') { should match(/storage-component.googleapis.com/) }
+  end
+
+  describe command("gcloud iam service-accounts list --project #{project_id} --format='json(email,disabled)'") do
```
Comment (morgante) [inline test/integration/vpc_sc_project/controls/vpc_sc_project.rb:50]: No need to add these redundant tests.

File: test/integration/vpc_sc_project/controls/vpc_sc_project.rb [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+# limitations under the License.
+
+project_id                         = attribute('project_id')
+project_number                     = attribute('project_number')
+service_account_email              = attribute('service_account_email')
+compute_service_account_email      = attribute('compute_service_account_email')
+group_email                        = attribute('group_email')
+group_name                         = attribute('group_name')
+
+control 'project-factory-vpc-sc-project' do
```
Comment (morgante) [inline test/integration/vpc_sc_project/controls/vpc_sc_project.rb:22]: We should add a test which validates presence in the perimeter.

File: test/setup/iam.tf [hunk 9]
```diff
@@ -57,6 +62,14 @@ resource "google_folder_iam_member" "int_test_folder" {
   member = "serviceAccount:${google_service_account.int_test.email}"
 }
 
+resource "google_organization_iam_member" "int_test_org" {
+  count = length(local.int_required_org_roles)
```
Comment (morgante) [inline test/setup/iam.tf:66]: Use for_each instead of count.

**Comment by bharathkkb** [comment]
@morgante I did it manually to unblock foundations, will send a PR soon.

File: test/integration/vpc_sc_project/controls/vpc_sc_project.rb (hunk 8 above)
Comment (bharathkkb) [reply test/integration/vpc_sc_project/controls/vpc_sc_project.rb:22]: Added

File: test/setup/iam.tf (hunk 9 above)
Comment (bharathkkb) [reply test/setup/iam.tf:66]: done

**Pull request merged**